hpcgap/lib/oper.g

```diff
@@ -1276,6 +1276,106 @@ BIND_GLOBAL( "DeclareAttribute", function ( arg )
 end );
 
 
+##############################################################################
+##
+##  This code adds thread-local attributes, which are used in HPC-GAP.
+##  These are not complete replacements for normal attributes.
+##  In particular, they are not filters, so cannot be used in method selection.
+##
+##  The main problem with thread-local attributes is that only a subset of the
+##  threads may have a value for a particular mutable attribute, but filters
+##  are not stored thread-locally. Therefore thread-local attributes are not
+##  a perfect replacement for traditional attributes.
+##
+##  The reason we have thread-local attributes is for code which uses
+##  mutable attribute as a general scratch-space, like StabChainMutable, which
+##  cannot be shared between threads.
+##
+##  Thread-local attributes where introduced to make existing
+##  GAP mutable attributes safe to use in a multi-threaded way, without
+##  significant code changes. If possible, no new thread local attributes should
+##  be added, and existing ones removed and cleaned up where possible.
+##
+##  Current Limitations:
+##
+##  In HPCGAP we disable the ImagesSet overload in lib/ghomperm.gi as it filters
+##  on HasStabChainMutable
+##
+##
+##  Thread local attributes must be mutable. The mutflag option is kept
+##  to keep function calls for DeclareAttribute and DeclareThreadLocalAttribute
+##  as close as possible
+##
+##  This function takes <name> and <filter> and <mutflag>, like DeclareAttribute,
+##  and creates functions called <name>, Set<name> and Has<name>, 
+##  like normal Attributes.
+##
+##  The main difference from normal attributes is that rather than calling
+##  InstallMethod on <name>, InstallMethod must be called on HPCGAP_TL<name>.
+##
+##  In non-HPC-GAP, DeclareThreadLocalAttribute behaves like
+##  DeclareAttribute, except it binds HPCGAP_TL<name> to <name>.
+##  This means InstallMethod can be called as normal, on both
+##  HPCGAP_TL<name> and <name>.
```

Comment:
That's of course correct, but should we really point it out? What purposes does that server? If code uses `DeclareThreadLocalAttribute`, then it should always use `InstallMethod` on `HPCGAP_TL<name>`, otherwise it'll not work in HPC-GAP, and then what's the point?

Reply:
Also, I am not really happy about the whole `HPCGAP_TL<name>`, that makes it counterintuitive to use TLS attributes. It's a nasty implementation detail...

I think we could hide it, though, by adding some kernel support for TLS attributes. Wouldn't be very difficult, I think. That said, I can totally understand not wanting to spend time on that, given that it is a feature we'd prefer to get rid of again, and which new code should ideally not use.  (I.e. I am OK with just keeping it as it is, and will just mutter and curse for a bit into my non-existent beard ;-)

Reply:
I warn you, unless you come up with something clever than me (very possible!) that adding thread-local attributes to the kernel seems very hard, because I can't see how to avoid having two flags for each objects, the thread local and the non-thread local. You have to update the non-thread local globally when you add new data, and then what happens if the additional of thread-local information causes a non-thread local attribute to be added?

There may be some clever route through this, but I tried several times and gave up each time.